packages/jest-message-util/src/index.ts

```diff
@@ -5,323 +5,294 @@
-    error.includes('ReferenceError: navigator is not defined')
-  ) {
-    return warnAboutWrongTestEnvironment(error, 'jsdom');
-  } else if (error.includes('.unref is not a function')) {
-    return warnAboutWrongTestEnvironment(error, 'node');
+function getWrongEnvironmentWarning(error: ErrorNode) {
+  if (error.message) {
+    const {message} = error.message;
+    let considerEnv: 'jsdom' | 'node' | undefined;
+    if (
+      (error.name?.name === 'ReferenceError' &&
+        message.includes('document is not defined')) ||
+      message.includes('window is not defined') ||
+      message.includes('navigator is not defined')
+    ) {
+      considerEnv = 'jsdom';
+    } else if (message.includes('.unref is not a function')) {
+      considerEnv = 'node';
+    }
+    if (considerEnv) {
+      return chalk.bold.red(
+        `The error below may be caused by using the wrong test environment, see ${chalk.dim.underline(
+          'https://jestjs.io/docs/configuration#testenvironment-string',
+        )}.\nConsider using the "${considerEnv}" test environment.`,
+      );
+    }
   }
 
-  return error;
-}
-
-function warnAboutWrongTestEnvironment(error: string, env: 'jsdom' | 'node') {
-  return (
-    chalk.bold.red(
-      `The error below may be caused by using the wrong test environment, see ${chalk.dim.underline(
-        'https://jestjs.io/docs/configuration#testenvironment-string',
-      )}.\nConsider using the "${env}" test environment.\n\n`,
-    ) + error
-  );
+  return undefined;
 }
 
-// ExecError is an error thrown outside of the test suite (not inside an `it` or
-// `before/after each` hooks). If it's thrown, none of the tests in the file
-// are executed.
-export const formatExecError = (
-  error: Error | TestResult.SerializableError | string | undefined,
-  config: StackTraceConfig,
-  options: StackTraceOptions,
-  testPath?: Path,
-  reuseMessage?: boolean,
-): string => {
-  if (!error || typeof error === 'number') {
-    error = new Error(`Expected an Error, but "${String(error)}" was thrown`);
-    error.stack = '';
-  }
-
-  let message, stack;
-
-  if (typeof error === 'string' || !error) {
-    error || (error = 'EMPTY ERROR');
-    message = '';
-    stack = error;
-  } else {
-    message = error.message;
-    stack =
-      typeof error.stack === 'string'
-        ? error.stack
-        : `thrown: ${prettyFormat(error, {maxDepth: 3})}`;
-  }
-
-  const separated = separateMessageFromStack(stack || '');
-  stack = separated.stack;
-
-  if (separated.message.includes(trim(message))) {
-    // Often stack trace already contains the duplicate of the message
-    message = separated.message;
+const cleanFrames = (
+  frames: Array<FrameNode> | undefined,
+  options: StackTraceOptions = {},
+): Array<FrameNode> => {
+  let frames_ = frames || [];
+  frames_ = frames_.filter(frame => !isInternalFrame(frame));
+
+  const firstFrame = frames_[0];
+
+  frames_ = frames_.filter(frame => !isJestInternalFrame(frame));
```

Comment:
This double filtering looks a little odd, but I prefer it since it means the callback can be completely stateless, where the previous callback needed to check whether it had ever previously returned true.